### PR TITLE
Remove runtime type-checking from enumerable types

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
This removes runtime type-checking for collections from Sorbet.

This type-checking is a frequent source of slowness, especially in surprising situations, and as per offline conversation we've decided to remove it.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by aisamanra** [comment]
One question I have: is it worth capturing the existing behavior of runtime type-checking for an enumerable in a helper method somewhere? I'm thinking that someone might want to have something like

```ruby
sig {params(xs: T::Array[Integer]).void}
def foo(xs)
  T.deep_type_assert!(xs, T::Array[Integer])
  # foo
end
```

which does the same thing as the current runtime verification, if people want to opt-in to that extra type-checking in some situations, but that it won't happen by default.

**Comment by aisamanra** [comment]
(Ah, I was lulled into a false sense of security by just seeing the LSP test failing in CI—I thought that meant it was the only one.)

I've just removed all those tests, since they were testing invariants we no longer expect to hold after this.

N.B. we should do internal comms when this lands, since it'll change expected runtime behavior.

File: gems/sorbet-runtime/test/types/props/decorator.rb [hunk 1]
```diff
@@ -172,18 +172,6 @@ class InnerStruct < T::Struct
     prop :the_hash, T::Hash[String, InnerStruct]
   end
 
-  describe 'validating prop values' do
```
Comment (jez) [inline gems/sorbet-runtime/test/types/props/decorator.rb:175]: Instead of deleting all these tests, can we convert them into tests that capture the new behavior?

**Comment by bbuchalter** [comment]
I recently uncovered this problem as well. Here's some performance metrics https://github.com/bbuchalter/sorbet_sig_performance

tl;dr 
When large hashes are fetched and type-checked at runtime, Sorbet has to run checks on every member in hash- which is more expensive than expected.
We have observed a linear degradation in performance when comparing the time it takes to read a hash with a sig vs. a hash with `.checked(:never)` ([doesn't run runtime checks](https://sorbet.org/docs/runtime#checked-whether-to-check-in-the-first-place)) specified.
For a `Hash[String, String]`:

| # of keys  |  Avg Method Calls/5 seconds (with sig) | Avg Method calls/5 seconds (sig never checked)  | Order of Magnitude of diff  |   |
|---|---|---|---|---|
| 5  | 454,777  |  2,308,000 | 1|
|  50 | 643,000  | 86,662,000  |  2| 
|  5000 | 7,020  |  96,780,000 | 4|

File: gems/sorbet-runtime/test/types/types.rb [hunk 2]
```diff
@@ -212,13 +212,10 @@ def subtype_of?(other)
...
 
-      it 'fails if an element of the array is the wrong type' do
+      it 'does not fail if an element of the array is the wrong type' do
         type = T::Array[Integer]
         value = [true]
-        msg = type.error_message_for_obj(value)
-        expected_error = "Expected type T::Array[Integer], " \
-                         "got T::Array[T::Boolean]"
-        assert_equal(expected_error, msg)
+        assert_nil(type.error_message_for_obj(value))
```
Comment (jez) [inline gems/sorbet-runtime/test/types/types.rb:218]: @aisamanra What does it mean for `error_message_for_obj` to be `nil`? Do we still show an "expected / actual" error message?

Comment (jez) [reply gems/sorbet-runtime/test/types/types.rb:218]: Oh I see, these are all saying that `error_message_for_obj` is `nil` because there is no error raised, because the only problem is that the element type is mismatched, not the class type.

File: gems/sorbet-runtime/lib/types/types/typed_enumerable.rb [hunk 3]
```diff
@@ -23,50 +23,7 @@ def name
...
-        return true
-      when Enumerator
-        # Enumerators can be unbounded: see `[:foo, :bar].cycle`
-        return true
-      when Range
-        # A nil beginning or a nil end does not provide any type information. That is, nil in a range represents
-        # boundlessness, it does not express a type. For example `(nil...nil)` is not a T::Range[NilClass], its a range
-        # of unknown types (T::Range[T.untyped]).
-        # Similarly, `(nil...1)` is not a `T::Range[T.nilable(Integer)]`, it's a boundless range of Integer.
-        (obj.begin.nil? || @type.valid?(obj.begin)) && (obj.end.nil? || @type.valid?(obj.end))
```
Comment (jez) [inline gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:57]: Maybe it doesn't make sense to do this, but `Range` always checks a constant number of elements (`.begin` and `.end`). It might be easy enough to keep this behavior, but I'm also fine removing it as you have here.

**Comment by jez** [comment]
@aisamanra It sounds like you were going to do some communications on Slack to announce this change.

Would you also mind updating the (fragment of a) page at https://sorbet.org/docs/stdlib-generics to mention maybe in a sentence or two that the type argument is only a hint to the static system, and is never checked at runtime (and maybe also why / the history?)

**Comment by aisamanra** [comment]
I went ahead and added the docs to in this PR. (I also added an extra sentence as long as I was already modifying those docs about the use of bare `Array` and the like: I was already there, after all!)

I'll also draft an internal email really quickly to do comms about this; we can discuss that offline.
